Adding support for Windows JNI build

koldat opened the pull request:
Initial support for building JNI library for Windows. Also some modification in Java unit tests to make it pass.

@yuslepukhin I was modifying CMakeLists.txt to add support for Windows JNI build. Can you check if changes are OK?

#703

yuslepukhin commented on CMakeLists.txt:
```diff
@@ -25,10 +25,20 @@
 cmake_minimum_required(VERSION 2.6)
 project(rocksdb)
 
+set(USE_JNI_DEFAULT 0)
```
thirdparty.inc is the place for such things

koldat replied:
But JNI is not third party library. The option there is just to instruct cmake to generate JNI project.  If you really do not like it, I can move it there.

adamretter commented on java/rocksjni/options.cc:
```diff
@@ -1180,7 +1179,7 @@ jbyte Java_org_rocksdb_Options_compactionStyle(
     JNIEnv* env, jobject jobj, jlong jhandle, jlong jmax_table_files_size) {
   reinterpret_cast<rocksdb::Options*>(jhandle)->compaction_options_fifo.max_table_files_size =
-    static_cast<long>(jmax_table_files_size);
+    static_cast<uint64_t>(jmax_table_files_size);
```
I am not a strong C++ developer, so excuse my ignorance, but why are we fixing this to 64bits unsigned?

koldat replied:
max_table_files_size is unit64_t. Casting to "long" took just lower 32 bits, so that setting and reading this variable was different when one used number bigger than 32bit (unit test).

adamretter replied:
Okay interesting. I wonder if this is a mistake that has been made elsewhere in the code base in the past?

koldat replied:
I do not know, but it was bug for Windows and I believe even for other platforms.

adamretter replied:
@koldat Okay it's something that I should check...

yuslepukhin replied:
ints and longs are 32-bit on windows in 64-bit builds

koldat replied:
I know, but casting int64 (jlong) to long is not correct. One will loose upper 32 bits which was correctly discovered by unit test. This change is fix for that.

adamretter commented on java/rocksjni/portal.h:
```diff
@@ -607,7 +607,7 @@ class WriteTypeJni {
 
     // Get the DELETE enum field of org.rocksdb.WBWIRocksIterator.WriteType
-    static jobject DELETE(JNIEnv* env) {
+    static jobject DELETE_(JNIEnv* env) {
```
Why have you renamed this?

koldat replied:
There is some macro DELETE on Windows platform (then we have compile error). I can also use undef if you see it better.

adamretter replied:
I think that might be better, otherwise we do not have a consistent naming for these enumerations.

adamretter commented on java/rocksjni/rocksjni.cc:
```diff
@@ -688,8 +693,8 @@ jint rocksdb_get_helper(
 
-  int cvalue_len = static_cast<int>(cvalue.size());
-  int length = std::min(jentry_value_len, cvalue_len);
+  jint cvalue_len = static_cast<int>(cvalue.size());
```
Why not `static_cast<jint>` or `reinterpret_cast<jint>` ?

koldat replied:
You are right, I forgot to update template argument. Important here was the final type. jint is int, but std:min failed on type conversion. I can update parameter to jint.

adamretter commented on java/src/main/java/org/rocksdb/NativeLibraryLoader.java:
```diff
@@ -75,37 +75,43 @@ public synchronized void loadLibrary(final String tmpDir) throws IOException {
+  File loadLibraryFromJarToTemp(final String tmpDir)
+          throws IOException {
+    final File temp;
+    if (tmpDir == null || tmpDir.equals("")) {
```
You could consider `tmpDir.isEmpty()` instead of `tmpDir.equals("")` as RocksJava is targeting Java 7 (but only using those features that are likely Android compatible).

koldat replied:
I was just extracting body of loadLibraryOfJar to reuse in unit test. I was not modifying the code. I can change it.

adamretter replied:
That would be nice, thanks :-)

adamretter commented on java/src/test/java/org/rocksdb/BackupableDBOptionsTest.java:
```diff
@@ -16,7 +17,7 @@
 public class BackupableDBOptionsTest {
 
-  private final static String ARBITRARY_PATH = "/tmp";
+  private final static String ARBITRARY_PATH = Environment.isWindows() ? System.getenv("TEMP") : "/tmp";
```
We should perhaps consider here instead: `System.getProperty("java.io.tmpdir")`

koldat replied:
I did not want to change behavior for other platforms so I fixed only Windows. Yes I can change it.

adamretter replied:
Great, I think such a change would improve the behaviour on all platforms.

adamretter commented on java/src/test/java/org/rocksdb/InfoLogLevelTest.java:
```diff
@@ -113,7 +114,7 @@ public void valueOf() {
    */
   private String getLogContentsWithoutHeader() throws IOException {
-    final String separator = System.getProperty("line.separator");
+    final String separator = Environment.isWindows() ? "\n" : System.getProperty("line.separator");
```
I don't see why you need this change? `System.getProperty("line.separator")` should return `"\n"` on Windows platforms. Have you tested that?

koldat replied:
No it returns "\r\n". RocksDB port is using native IO functions so that it does not convert "\n" -> "\r\n". Then unit tests fails.

adamretter replied:
Ahhh I see, then perhaps we should just fix it to `\n`?

koldat replied:
OSX has it different... "\r" only.

adamretter replied:
Ah of course! Sorry, i'm a little tired ;-)

koldat commented:
@yuslepukhin @adamretter I pushed changes that should address code review issues.

adamretter commented:
@koldat Looks good to me. @yhchiang are we still using Phabricator for merging changes?

koldat commented:
@adamretter I changed a way of how to reference int_backup_ids. When vector is empty it cannot be dereferenced.

adamretter commented:
@koldat Okay cool, I can't really comment though as my C++ is not strong enough to know what the implications of that change are.

yuslepukhin commented on CMakeLists.txt:
```diff
@@ -287,6 +288,8 @@ set_target_properties(rocksdb${ARTIFACT_SUFFIX} PROPERTIES COMPILE_FLAGS "-DROCK
 add_dependencies(rocksdb${ARTIFACT_SUFFIX} GenerateBuildVersion)
 target_link_libraries(rocksdb${ARTIFACT_SUFFIX} ${LIBS})
 
+include(${CMAKE_CURRENT_SOURCE_DIR}/java/jni.inc)
```
@koldat  The main point here is not to make the main file cluttered. So a separate file and a separate folder is  a good idea. Since many do not build java here is additional ideas.

1. Include this file optionally based on the command line option OR
1.  Since you say it is a separate project you can instead of a include file write it in a separate CMakeLists.txt in that subfolder and conditionally add it as a separate project based on the command line. Like here `rocksdb/third-party/gtest-1.7.0/fused-src/gtest using add_subdirectory

koldat replied:
1. Ok I can do that
2. I am reusing setup, variables, targets from main cmake. That us why I used include. I did not want to loose changes that anyone will do in main cmake. JNI is just little bit different version of rocksdb.dll. I do not think it is different project, but also not third party. Include is easiest way how to achieve that.

koldat replied:
Ad1) JNI is already enabled in conditional way. I just have condition inside like in third party. Do you really want to move it? It should be same as for other libraries (condition is inside third-party.Inc)

yuslepukhin replied:
1. Conditional include would avoid file processing alltogether

1. A separate project would inherit the settings from the main file. Check the gtest out. There all is done is add_library. Either way you do not have to change much.

yuslepukhin replied:
@koldat Pls, also make sure things work in release  build. So 4 permutations 1,2) Debug with and w/o Java 3,4) same with Release

koldat commented:
@yuslepukhin I changed JNI to subproject (like gtest) and rebased branch so it should pass the build.

siying commented on java/rocksjni/backupablejni.cc:
```diff
@@ -105,10 +105,9 @@ jintArray Java_org_rocksdb_BackupableDB_getCorruptedBackups(
-  for (std::vector<rocksdb::BackupID>::size_type i = 0;
-      i != kIdSize; i++) {
-    int_backup_ids[i] = backup_ids[i];
+  std::auto_ptr<jint> int_backup_ids(new jint[kIdSize]);
```
We don't use auto_ptr in our code base. Can you explain why you don't simply use std::vector<jint>?

koldat commented:
@siying I changed int_backup_ids back to vector. The problem was dereferencing empty vector. I used data() method that was added in C++11, that should work even for empty vectors. I also squashed changesets to one, rebased to latest and fixed one test that was not passing with release DLLs.

yhchiang commented:
>  @yhchiang are we still using Phabricator for merging changes?

For bigger change, yes it is suggested to use Phabricator as it's easier to review and discuss.

siying commented:
@adamretter @yuslepukhin is it good to merge?

adamretter commented:
@siying All good from me
